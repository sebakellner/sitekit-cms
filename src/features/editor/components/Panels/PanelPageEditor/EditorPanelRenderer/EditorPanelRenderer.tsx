import { PanelBoxCollapsible } from '@components/cms/ui'
import EditorRenderer from '../EditorRenderer/EditorRenderer'
import type { EditorPanelRendererProps } from './EditorPanelRenderer.types'

const EditorPanelRenderer = ({
  meta,
  values,
  onChange,
}: EditorPanelRendererProps) => {
  if (!meta.panels) return null

  return (
    <>
      {meta.panels.map(({ id, title, fields }) => (
        <PanelBoxCollapsible key={id} title={title}>
          {fields.map((fieldKey) => {
            const propConfig = meta.props[fieldKey]

            if (!propConfig) return null

            const configWithTitle = {
              ...propConfig,
              title: propConfig.title ?? fieldKey,
            }

            return (
              <EditorRenderer
                key={fieldKey}
                propConfig={configWithTitle}
                value={values[fieldKey]}
                onChange={(val) => onChange(fieldKey, val)}
              />
            )
          })}
        </PanelBoxCollapsible>
      ))}
    </>
  )
}

export default EditorPanelRenderer
