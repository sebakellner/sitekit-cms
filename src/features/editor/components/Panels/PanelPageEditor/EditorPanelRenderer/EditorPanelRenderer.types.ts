import type { GroupedPanel } from '@features/editor/hooks/useGroupedPropsPanels'

export type EditorPanelRendererProps = {
  groupedPanels: GroupedPanel[]
  onChange: (key: string, value: unknown) => void
}
