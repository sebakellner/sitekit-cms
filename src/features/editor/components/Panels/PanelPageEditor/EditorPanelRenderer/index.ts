export { default as EditorPanelRenderer } from './EditorPanelRenderer'
