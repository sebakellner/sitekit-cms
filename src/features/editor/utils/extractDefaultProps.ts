import { ComponentMetaSchema } from '@components/site/schemas/componentMeta.schema'
import type { ComponentMeta } from '@components/site/types'

function isDefaultProp(val: unknown): val is { default: unknown } {
  return typeof val === 'object' && val !== null && 'default' in val
}

export function extractDefaultProps(
  props: ComponentMeta['props'] | undefined,
  fallback: unknown = undefined
): Record<string, unknown> {
  if (!props) return {}
  const result = ComponentMetaSchema.shape.props.safeParse(props)
  if (!result.success) {
    console.error('Invalid props schema:', result.error)
    return {}
  }
  const validProps = result.data
  return Object.fromEntries(
    Object.entries(validProps).map(([key, val]) => [
      key,
      isDefaultProp(val) ? val.default : fallback,
    ])
  )
}
