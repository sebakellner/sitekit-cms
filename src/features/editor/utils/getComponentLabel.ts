import React from 'react'

export function getComponentLabel(
  children: React.ReactNode,
  fallback = 'Component'
): string {
  if (React.isValidElement(children)) {
    const type = children.type
    if (typeof type === 'function' && type.name) return type.name
  }
  return fallback
}
