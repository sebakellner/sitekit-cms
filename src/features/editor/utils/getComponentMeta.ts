import { ComponentMetaSchema } from '@components/site/schemas/componentMeta.schema'

type ModuleWithDefault = { default: unknown }

export function getComponentMeta(module: unknown) {
  if (
    module &&
    typeof module === 'object' &&
    'default' in module &&
    (module as ModuleWithDefault).default &&
    typeof (module as ModuleWithDefault).default === 'object'
  ) {
    const def = (module as ModuleWithDefault).default
    const result = ComponentMetaSchema.safeParse(def)
    if (result.success) {
      return result.data
    }
  }
  return { name: '', props: {} }
}
