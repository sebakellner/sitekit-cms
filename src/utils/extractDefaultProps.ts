import type { ComponentMeta } from '@components/site/types'

export function extractDefaultProps(
  props: ComponentMeta['props'] | undefined
): Record<string, unknown> {
  if (!props) return {}
  return Object.fromEntries(
    Object.entries(props).map(([key, val]) => [
      key,
      (val as { default: unknown }).default,
    ])
  )
}
