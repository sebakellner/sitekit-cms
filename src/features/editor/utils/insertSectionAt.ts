import { componentsRegistry } from '@registry/componentRegistry'
import { v4 as uuidv4 } from 'uuid'

import type { ComponentMeta } from '@components/site/types'
import { extractDefaultProps } from '@src/features/editor/utils/extractDefaultProps'
import type { Section } from '../schemas/section.schema'

export async function insertSectionAtUtil(
  sections: Section[],
  setSections: (sections: Section[]) => void,
  name: string,
  index: number
) {
  const loader = componentsRegistry[name as keyof typeof componentsRegistry]
  if (!loader) {
    console.warn(`Component "${name}" not found in registry.`)
    return
  }

  const metaModule = await loader()

  const meta = metaModule.default as ComponentMeta

  const defaultProps = extractDefaultProps(meta?.props)

  const newSection: Section = {
    id: uuidv4(),
    name,
    props: defaultProps,
  }

  const updatedSections = [
    ...sections.slice(0, index),
    newSection,
    ...sections.slice(index),
  ]

  setSections(updatedSections)
}
