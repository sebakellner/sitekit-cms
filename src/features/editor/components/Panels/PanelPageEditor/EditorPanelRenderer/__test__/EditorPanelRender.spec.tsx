import { expect, describe, test } from 'vitest'
import EditorPanelRenderer from '../EditorPanelRenderer'
import type { EditorPanelRendererProps } from '../EditorPanelRenderer.types'
import { render } from 'test/utils/render'
import { fireEvent, screen } from '@testing-library/react'

const setup = (overrides: Partial<EditorPanelRendererProps> = {}) => {
  const defaultProps: EditorPanelRendererProps = {
    groupedPanels: mockGroupedPanels,
    onChange: () => {},
  }
  const props = { ...defaultProps, ...overrides }
  return <EditorPanelRenderer {...props} />
}

describe('EditorPanelRenderer', () => {
  test('should render the component with the correct metadata', () => {
    const { container } = render(setup())
    expect(container).toMatchSnapshot()
  })

  test('should render correctly', () => {
    render(setup({}))
    expect(screen.getByText('Panel 1')).toBeInTheDocument()
  })

  test('should call onChange when a field value changes', () => {
    const mockOnChange = vi.fn()
    render(setup({ onChange: mockOnChange }))

    const input = screen.getByDisplayValue('Option 1')

    fireEvent.change(input, { target: { value: 'option2' } })

    expect(mockOnChange).toHaveBeenLastCalledWith('prop1', 'option2')
  })

  test('should handle undefined groupedPanels', () => {
    render(setup({ groupedPanels: undefined }))
    expect(screen.getByText('No properties available')).toBeInTheDocument()
  })
})
