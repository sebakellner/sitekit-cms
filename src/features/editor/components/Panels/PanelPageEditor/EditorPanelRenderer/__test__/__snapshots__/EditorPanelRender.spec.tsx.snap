// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditorPanelRenderer > should render the component with the correct metadata 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="css-8ei852"
    data-testid="panel-box"
  >
    <div
      class="css-pg61ge"
    >
      <h6
        class="chakra-heading css-1na86dk"
      >
        Panel 1
      </h6>
      <button
        class="chakra-button css-ljrtzn"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-down"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          style="transform: none; transition: transform 0.2s;"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </button>
    </div>
    <div
      class="css-12wustm"
    >
      <div
        class="css-n33ttv"
      >
        <div
          class="chakra-field__root css-1anu6qw"
          data-part="root"
          data-scope="field"
          id="field::«r0»"
          orientation="horizontal"
          role="group"
        >
          <label
            class="chakra-field__label css-84rnco"
            data-part="label"
            data-scope="field"
            for="«r0»"
            id="field::«r0»::label"
          >
            Property 1
          </label>
          <div
            class="chakra-native-select__root css-8sp2e0"
          >
            <select
              class="chakra-native-select__field css-jwpnxr"
              data-part="select"
              data-scope="field"
              id="«r0»"
            >
              <option
                value="option1"
              >
                Option 1
              </option>
              <option
                value="option2"
              >
                Option 2
              </option>
            </select>
            <div
              class="chakra-native-select__indicator css-hydxzs"
            >
              <svg
                class="css-s3mb0o"
                viewBox="0 0 24 24"
              >
                <path
                  d="m6 9 6 6 6-6"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-n33ttv"
      >
        <div
          class="chakra-field__root css-1anu6qw"
          data-part="root"
          data-scope="field"
          id="field::«r1»"
          orientation="horizontal"
          role="group"
        >
          <label
            class="chakra-field__label css-84rnco"
            data-part="label"
            data-scope="field"
            for="«r1»"
            id="field::«r1»::label"
          >
            Property 2
          </label>
          <input
            class="chakra-input css-1hens3x"
            data-part="input"
            data-scope="field"
            id="«r1»"
            value="new value"
          />
        </div>
      </div>
      <div
        class="css-cdgtm1"
      >
        <div
          class="chakra-field__root css-1anu6qw"
          data-part="root"
          data-scope="field"
          id="field::«r2»"
          orientation="horizontal"
          role="group"
        >
          <label
            class="chakra-field__label css-84rnco"
            data-part="label"
            data-scope="field"
            for="«r2»"
            id="field::«r2»::label"
          >
            Property 3
          </label>
          <div
            class="colorPicker__root css-zxnr2f"
            data-part="root"
            data-scope="color-picker"
            dir="ltr"
            id="color-picker:«r3»"
            style="--value: rgba(255, 0, 0, 1);"
          >
            <input
              id="color-picker:«r3»:hidden-input"
              style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap; word-wrap: normal;"
              tabindex="-1"
              type="text"
              value="rgba(255, 0, 0, 1)"
            />
            <div
              class="colorPicker__control css-1q6y153"
              data-part="control"
              data-scope="color-picker"
              data-state="closed"
              dir="ltr"
              id="color-picker:«r3»:control"
            >
              <input
                aria-label="hex"
                autocomplete="off"
                class="colorPicker__channelInput css-1aemz6e"
                data-channel="hex"
                data-part="channel-input"
                data-scope="color-picker"
                dir="ltr"
                max="16777215"
                min="0"
                spellcheck="false"
                step="1"
                style="appearance: none;"
                type="text"
                value="#FF0000"
              />
              <button
                aria-controls="color-picker:«r3»:content"
                aria-label="select color. current color is rgba(255, 0, 0, 1)"
                aria-labelledby="field::«r2»::label"
                class="colorPicker__trigger css-10mv649"
                data-part="trigger"
                data-scope="color-picker"
                data-state="closed"
                dir="ltr"
                id="color-picker:«r3»:trigger"
                style="position: relative;"
                type="button"
              >
                <div
                  class="colorPicker__swatch css-lkq982"
                  data-part="swatch"
                  data-scope="color-picker"
                  data-state="checked"
                  data-value="#FF0000"
                  dir="ltr"
                  style="--color: rgba(255, 0, 0, 1); position: relative; background: rgb(255, 0, 0);"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
